projects/driver_controls/src/throttle.c

```diff
@@ -0,0 +1,159 @@
+static uint16_t prv_scale_reading(int16_t reading, int16_t max, int16_t min) {
+  return (1 << 12) * (reading - min) / (max - min);
+}
+
+static uint16_t prv_get_numerator(int16_t reading, ThrottleZone zone, ThrottleChannel channel,
+                                  ThrottleStorage *storage) {
+  if (channel == THROTTLE_CHANNEL_MAIN) {
+    return prv_scale_reading(
+        reading, storage->calibration_data->zone_thresholds_main[zone][THROTTLE_THRESH_MAX],
+        storage->calibration_data->zone_thresholds_main[zone][THROTTLE_THRESH_MIN]);
+  } else {
+    return prv_scale_reading(
+        reading, storage->calibration_data->zone_thresholds_secondary[zone][THROTTLE_THRESH_MAX],
+        storage->calibration_data->zone_thresholds_secondary[zone][THROTTLE_THRESH_MIN]);
+  }
+}
+
+static bool prv_reading_within_zone(int16_t reading_main, ThrottleZone zone,
+                                    ThrottleStorage *storage) {
+  return (
+      (reading_main < storage->calibration_data->zone_thresholds_main[zone][THROTTLE_THRESH_MAX]) &&
+      (reading_main >= storage->calibration_data->zone_thresholds_main[zone][THROTTLE_THRESH_MIN]));
+}
+
+// Returns true if the channels match their supposed relationship.
+static bool prv_channels_synced(ThrottleStorage *storage, int16_t reading_main,
+                                int16_t reading_secondary) {
+  uint16_t numerator_main =
+      prv_get_numerator(reading_main, THROTTLE_ZONE_ALL, THROTTLE_CHANNEL_MAIN, storage);
+
+  uint16_t numerator_secondary =
+      prv_get_numerator(reading_secondary, THROTTLE_ZONE_ALL, THROTTLE_CHANNEL_SECONDARY, storage);
+
+  return (abs(numerator_main - numerator_secondary)) <=
+         storage->calibration_data->channel_readings_tolerance;
+}
+
+// This callback is called whenever a conversion is done. It sets the flags
+// so that we know conversions are happening.
+static void prv_flag_update_callback(Ads1015Channel channel, void *context) {
+  if (context == NULL) {
+    return;
+  }
+  ThrottleStorage *storage = context;
+  storage->reading_updated_flag = true;
+}
+
+// The periodic callback which checks if readings are up to date and channels are in sync.
+// If they are, the event corresponding to the pedals position is raised.
+// If not, a pedal timout event is raised.
+static void prv_raise_event_timer_callback(SoftTimerID timer_id, void *context) {
+  ThrottleStorage *storage = context;
+  int16_t reading_main = ADS1015_READ_UNSUCCESSFUL;
+  int16_t reading_secondary = ADS1015_READ_UNSUCCESSFUL;
+
+  ads1015_read_raw(storage->pedal_ads1015_storage, storage->channel_main, &reading_main);
+
+  ads1015_read_raw(storage->pedal_ads1015_storage, storage->channel_secondary, &reading_secondary);
+
+  InputEvent event = NUM_INPUT_EVENTS;
+  ThrottleZone zone = NUM_THROTTLE_ZONES;
+
+  if (storage->reading_updated_flag &&
+      prv_channels_synced(storage, reading_main, reading_secondary) &&
```

Comment:
If you already have all three zones, why do you need an "all" threshold pair? You could do the channel sync check in your zone check.